### ISSUE TITLE
Support for aliases of commands with multiple names

### ISSUE THREAD

**Issue by oxguy3** [tldr-pages/tldr#787]
It'd be really useful if there were a way to note that one command is an alias of another. Many tools are distributed under multiple names, and the TLDR client ought to figure this out instead of leaving it to the user to decipher. Examples:
- A lot of GNU utilities get prefixed with a "g" when installed on systems like OS X.
- `python2` and `python3` are typically aliases of `python`, and so both those pages ought to redirect to the python.md page. Same applies to the executables of many other languages.
- In many cases, two different commands will actually just be slight variations of the same underlying tool, and don't really need separate pages. `cowthink` is just a variant of `cowsay`, `7z`/`7za`/`7zr` are all minor variants on the exact same utility, etc. In general, if two commands share the same man page, they probably ought to share the same tldr page as well.

Here's an example of how this might be implemented. The OS X tool "gdate" is really just the GNU tool "date". So _pages/osx/gdate.md_ could perhaps look like this:

```
# gdate

> REDIRECT: common/date
```

TLDR clients that recognize this alias syntax could automatically load the common/date page (obviously there'd have to be delay between announcing this and implementing it so clients could update to support it).

Implementation note: It'd be ideal if the build server checked that all redirects are valid, and that there are no redirect chains (i.e. you shouldn't be able redirect to another redirect. That would just create unnecessary complexity, and it'd also cause lots of extraneous HTTP requests for any client that queries the live repo instead of using a cached local copy.)

**Comment by jedahan** [comment]
why not just use symlinks?

**Comment by oxguy3** [comment]
Huh, guess I forgot git supported those. Will definitely need to make sure that it doesn't break anything (I wonder how GitHub handles it when, say, a JS-based client tries to download a symlinked file), but yeah I guess that might work better.

**Comment by waldyrious** [comment]
Do they work seamlessly on all platforms?

**Comment by jedahan** [comment]
As far as I can tell, symlinks work fine on osx, linux, free/openbsd. They do not seem to work on windows (barring ugly post-download hacks), but almost all of these commands require some sort of *nix environment, and are meant for quick lookup. So I think its OK.

The browser-as-client question is way more interesting, and we should find a way to make that work well. But I think that might be a peculiarity of that client, that can be worked around - it would indeed be nice if only the JS client needed changing to support symlinks, rather than us coming up with our own format and requiring all clients to implement it.

**Comment by ostera** [comment]
tldr: already supported by the web client. You have my :+1: @oxguy3 

From a very simple test: https://api.github.com/repos/ostera/symlink-test/contents/

Yields:

```json
[
  {
    "name": "linked.md",
    "path": "linked.md",
    "sha": "828940fad27634bc54476cd0e4b1cc8c4a3838d8",
    "size": 6,
    "url": "https://api.github.com/repos/ostera/symlink-test/contents/linked.md?ref=master",
    "html_url": "https://github.com/ostera/symlink-test/blob/master/linked.md",
    "git_url": "https://api.github.com/repos/ostera/symlink-test/git/blobs/828940fad27634bc54476cd0e4b1cc8c4a3838d8",
    "download_url": "https://raw.githubusercontent.com/ostera/symlink-test/master/linked.md",
    "type": "file",
    "_links": {
      "self": "https://api.github.com/repos/ostera/symlink-test/contents/linked.md?ref=master",
      "git": "https://api.github.com/repos/ostera/symlink-test/git/blobs/828940fad27634bc54476cd0e4b1cc8c4a3838d8",
      "html": "https://github.com/ostera/symlink-test/blob/master/linked.md"
    }
  },
  {
    "name": "tee.md",
    "path": "tee.md",
    "sha": "bc9070a122d06709415346f14a56a58c9dd2f9ad",
    "size": 257,
    "url": "https://api.github.com/repos/ostera/symlink-test/contents/tee.md?ref=master",
    "html_url": "https://github.com/ostera/symlink-test/blob/master/tee.md",
    "git_url": "https://api.github.com/repos/ostera/symlink-test/git/blobs/bc9070a122d06709415346f14a56a58c9dd2f9ad",
    "download_url": "https://raw.githubusercontent.com/ostera/symlink-test/master/tee.md",
    "type": "file",
    "_links": {
      "self": "https://api.github.com/repos/ostera/symlink-test/contents/tee.md?ref=master",
      "git": "https://api.github.com/repos/ostera/symlink-test/git/blobs/bc9070a122d06709415346f14a56a58c9dd2f9ad",
      "html": "https://github.com/ostera/symlink-test/blob/master/tee.md"
    }
  }
]
```

Whereas going straight to the file (which is the way https://ostera.github.io/tldr.jsx does it): https://api.github.com/repos/ostera/symlink-test/contents/linked.md

Resolves the link automatically:

```json
{
  "name": "linked.md",
  "path": "linked.md",
  "sha": "828940fad27634bc54476cd0e4b1cc8c4a3838d8",
  "size": 257,
  "url": "https://api.github.com/repos/ostera/symlink-test/contents/linked.md?ref=master",
  "html_url": "https://github.com/ostera/symlink-test/blob/master/linked.md",
  "git_url": "https://api.github.com/repos/ostera/symlink-test/git/blobs/828940fad27634bc54476cd0e4b1cc8c4a3838d8",
  "download_url": "https://raw.githubusercontent.com/ostera/symlink-test/master/linked.md",
  "type": "file",
  "content": "IyB0ZWUKCj4gUmVhZCBmcm9tIHN0YW5kYXJkIGlucHV0IGFuZCB3cml0ZSB0\nbyBzdGFuZGFyZCBvdXRwdXQgYW5kIGZpbGVzLgoKLSBDb3B5IHN0YW5kYXJk\nIGlucHV0IHRvIGVhY2ggRklMRSwgYW5kIGFsc28gdG8gc3RhbmRhcmQgb3V0\ncHV0OgoKYGVjaG8gImV4YW1wbGUiIHwgdGVlIHt7RklMRX19YAoKLSBBcHBl\nbmQgdG8gdGhlIGdpdmVuIEZJTEVzLCBkbyBub3Qgb3ZlcndyaXRlOgoKYGVj\naG8gImV4YW1wbGUiIHwgdGVlIC1hIHt7RklMRX19YAo=\n",
  "encoding": "base64",
  "_links": {
    "self": "https://api.github.com/repos/ostera/symlink-test/contents/linked.md?ref=master",
    "git": "https://api.github.com/repos/ostera/symlink-test/git/blobs/828940fad27634bc54476cd0e4b1cc8c4a3838d8",
    "html": "https://github.com/ostera/symlink-test/blob/master/linked.md"
  }
}
```

**Comment by igorshubovych** [comment]
Symlink are not supported on Windows.

**Comment by ostera** [comment]
Actually some sort of symlinks are supported on Windows 2000 and onwards. I think Vista already supports symbolic links – considering the paths will be `../<platform>/<name>.md` it might work.

Besides, `cygwin` does support (and workarounds) symlinks too, with our without native NTFS support enabled.

Worth giving a shot.

**Comment by igorshubovych** [comment]
I know that this is ongoing issue for Docker, VirtualBox and some other cross-platform software.

**Comment by ostera** [comment]
Well, not sure how `docker` is having issues given that it just can't run on Windows at all (you'd have to run a virtual machine with a kernel with `cgroups` support to run `lxc`, and that is the Linux kernel). 

The problem with symlinking usually happens when you symlink from the guest to the host machine, when the host machine is running Windows.

Even if it didn't work, I think it will be degraded performance rather than a failure – for the case of a command with the exact same name in all platforms, on clients that still fallback to `common` if the command isn't found under the `windows` folder, then it will still show the right page; for the case of a command with a different name as `gdate`, the person running a cli tldr-client on windows and only under the very specific scenario of not using a linux-flavored environment but instead the built-in shell/terminal combo, and if they have a version of windows that doesn't support symlinks at all (prior to win2k I believe), _then_ they wouldn't get the page.

Now that would imply that the index we build may include entries that are not accessible to windows users that don't use anything like cywgin, but we either try to include those users (who from the stats I have on tldr.jsx are about 30% of the traffic, from where my gut tells me at least half use a VM, and a quarter uses cygwin at least) and we build our own redirection *OR* we go for the bigger chunk of our users and try to make their lives way easier by relying on symlinks for aliases.

Some food for thought I guess :)

**Comment by ostera** [comment]
ping. @oxguy3 @waldyrious @igorshubovych 

What do we want to do with this?

**Comment by waldyrious** [comment]
I'll defer the decision to those with more technical background than me :) whatever you agree makes the most sense is OK with me.

**Comment by waldyrious** [comment]
@ostera @agnivade @tldr-pages/content shall we give this a shot? If this works the benefits are great, and if it doesn't we won't be any worse than we're right now, so why not? :)

**Comment by ostera** [comment]
As I said before, the web client already supports this, so I'm +💯 ! Let's start symlinkin'!

**Comment by waldyrious** [comment]
Do I get it right -- we just need to create a symlink, commit it to git and everything should Just Work™?

**Comment by ostera** [comment]
I think so, yes. We can do a small pr and try it out

**Comment by agnivade** [comment]
Lets do this !

**Comment by waldyrious** [comment]
I tried this locally (using the node client):

```
$ cd pages/common/
$ ln -s crontab.md cron.md
$ tldr cron
Page not found
Try updating with "tldr --update", or submit a pull request to:
https://github.com/tldr-pages/tldr
$ ls -l cron.md 
lrwxr-xr-x  1 waldyrious  staff  10 Oct  2 15:48 cron.md -> crontab.md
$ ln -sf crontab.md ./cron.md
$ tldr cron
Page not found
Try updating with "tldr --update", or submit a pull request to:
https://github.com/tldr-pages/tldr
$ ls -l cron.md 
lrwxr-xr-x  1 waldyrious  staff  10 Oct  2 15:49 cron.md -> crontab.md
```

Any hints on what I could be doing wrong?

**Comment by sbrl** [comment]
@waldyrious What happens if you edit the original file and then do a `tldr crontab`? Does it update?

**Comment by waldyrious** [comment]
@sbrl now it doesn't print anything -- no "Page not found" message, no anything :/

**Comment by sbrl** [comment]
Ouch. Uh oh!

**Comment by zdroid** [comment]
@waldyrious I think I solved this. You need to make a permanent link using `ln` instead of soft link using `ln -s`.

```
火 ~/dev/tldr ln pages/common/crontab.md pages/common/cron.md
火 ~/dev/tldr tldr --render pages/common/cron.md 

  crontab

  Schedule cron jobs to run on a time interval for the current user.
  Job definition format: "(min) (hour) (day_of_month) (month) (day_of_week) command_to_execute".

  - Edit the crontab file for the current user:
    crontab -e

  - Edit the crontab file for a specific user:
    sudo crontab -e -u user

  - View a list of existing cron jobs for current user:
    crontab -l

  - Remove all cron jobs for the current user:
    crontab -r

  - Sample job which runs at 10:00 every day (* means any value):
    0 10 * * * command_to_execute

  - Sample job which runs every minute on the 3rd of April:
    * * 3 Apr * command_to_execute

  - Sample job which runs a certain script at 02:30 every Friday:
    30 2 * * Fri /absolute/path/to/script.sh
```

Time to symlink, I guess?

**Comment by waldyrious** [comment]
@zdroid funny you mention this now, just the other day I thought about this after [noticing](https://github.com/Homebrew/homebrew-core/pull/53043) how Homebrew does it. They do use regular symbolic links, though, so I'm not sure if this is something that should be solved on the clients' side. Do symbolic links fail for you the same way as they [did for me](https://github.com/tldr-pages/tldr/issues/787#issuecomment-250975326)?

And if so — @agnivade do you know if fixing the behavior with symlinks is possible / desirable for the node client?

Otherwise, we could move ahead with hard links, but I have no idea if that has any downsides compared to symbolic links. Does anyone know? If there are no downsides and it works with existing clients (and assuming that symbolic links don't), I guess we could move ahead, yeah.

**Comment by zdroid** [comment]
Clients don't seem to work with symlinks, so hard links are a solution for that. Although an alias file (maybe even as `.json`) may not be that bad after all.

**Comment by sbrl** [comment]
Are you sure that Git supports hard links? I would have thought that it would treated 2 hard-linked files as 2 separate files? Hard links require explicit support in programs to not be treated as 2 separate files.

(also, as much as I might hate to ask the question, does Windows support hard links? I know it supports directory junctions - but I'm not sure about anything else)

**Comment by zdroid** [comment]
I tried it locally and both hard and soft links were tracked the same way. But (potentially outdated) links like https://stackoverflow.com/questions/3729278/git-and-hard-links say it doesn't work.

On the other side, wouldn't a file or directory with aliases resolve the issue much more efficiently? (it would remove the need to make symlinks inside non-English pages)

**Comment by sbrl** [comment]
Ah, I see @zdroid. So it was implemented at some point - so it would work for newer git clients, but not for older ones.

It also wouldn't work for the archive itself that we publish, but that's not a particular issue because don't expect to be making changes there.

That's a great idea @zdroid! Do you mean something like a special redirect syntax or something?

e.g. if `common/foo` is the master copy, then `common/bar` might look something like this?

```markdown
# REDIRECT common/foo
```

Then non-supporting clients would handle it fine, but we'd get progressive enhancement for client  that do support it. This would necessitate an update to the client spec, but that's not a big deal.

**Comment by zdroid** [comment]
Something like that. It should work well enough, so we can try it.

**Comment by marchersimon** [comment]
Could we consider this issue as resolved, since we already introduced alias pages?

**Comment by pixelcmtd** [comment]
I think we should.

**Comment by sbrl** [comment]
Indeed, I think we can close this issue now.
